Sources/MongoSwift/ClientSession.swift

```diff
@@ -157,11 +122,14 @@ public final class ClientSession {
     /// Destroy the underlying `mongoc_client_session_t` and set this session to inactive.
     /// Does nothing if this session is already inactive.
     internal func end() {
-        if case let .active(session, connection) = self.state {
-            mongoc_client_session_destroy(session)
-            self.client.connectionPool.checkIn(connection)
-            self.state = .inactive
+        guard self.active else {
+            return
         }
+        let operation = EndSessionOperation(session: self)
+        // since `end()` is called from `deinit` and a `defer` block in `withSession`, we can't throw here. however,
+        // in practice this will never throw anyway. the branch taken in `executeOperation` doesn't throw, and neither
+        // does `EndSessionOperation`'s execute method.
+        try? self.client.executeOperation(operation)
```

Comment:
exact details are to be worked out in async design, but idea is that we can execute this asynchronously if the client used to create this session was an async client